opa-manager-operator/tests/integration/test_charm.py

```diff
@@ -1,115 +1,144 @@
+    )
+    await model.block_until(
+        lambda: "gatekeeper-controller-manager" in model.applications, timeout=60
+    )
+    await model.wait_for_idle(status="active")
 
 
 @pytest.mark.abort_on_fail
-async def test_build_and_deploy(ops_test):
-    charm = await ops_test.build_charm(".")
-    # Due to: https://github.com/juju/python-libjuju/issues/515
-    #  We have to use the k8s API to wait, we cannot use:
-    # ops_test.model.applications['gatekeeper'].units[0].workload_status
-    model_name = ops_test.model_name
-    role_binding_file = Path("/tmp/k8s-rolebinding.yaml")
-    with open("docs/gatekeeper-rb.yaml.template", "r") as fh:
-        template = Template(fh.read())
-        role_binding_file.write_text(
-            template.render(
-                service_account_user=f"system:serviceaccount:{model_name}:opa-manager-test-operator"
+async def test_apply_policy(client):
+    # Create a template and a constraint
+    load_in_cluster_generic_resources(client)
+    policy = codecs.load_all_yaml(
+        Path("docs/policy-example.yaml").read_text(), create_resources_for_crds=True
+    )[0]
+    ConstraintTemplate = get_generic_resource(
+        "templates.gatekeeper.sh/v1", "ConstraintTemplate"
+    )
+    client.create(policy)
+    # Wait for the template crd to be created from gatekeeper
+    client.wait(
+        CustomResourceDefinition,
+        "k8srequiredlabels.constraints.gatekeeper.sh",
+        for_conditions=("Established",),
+    )
+    load_in_cluster_generic_resources(client)
+    assert client.get(ConstraintTemplate, policy.metadata.name)
+
+    constraint = codecs.load_all_yaml(
+        Path("docs/policy-spec-example.yaml").read_text(),
+        create_resources_for_crds=True,
+    )[0]
+    client.create(constraint)
+
+    load_in_cluster_generic_resources(client)
+    Constraint = get_generic_resource(
+        "constraints.gatekeeper.sh/v1beta1", policy.spec["crd"]["spec"]["names"]["kind"]
+    )
+    # Verify that the constraint was created
+    assert client.get(Constraint, constraint.metadata.name)
+
+    ns_name = f"test-ns-{random.randint(1, 99999)}"
+    # Verify that the policy is enforced
+    for _ in range(10):
+        with pytest.raises(ApiError) as e:
+            client.create(Namespace(metadata=ObjectMeta(name=ns_name)))
+            log.info("Namespace was created, the policy was not enforced")
+            # The policy was not enforce, clean the workspace
+            client.delete(Namespace, ns_name)
+
+        err_msg = e.value.response.json()["message"]
+        if e.value.response.status_code == 500:
+            # the constraint has not been created yet, retry
+            assert err_msg.startswith(
+                "Internal error occurred: failed calling webhook "
+                '"check-ignore-label.gatekeeper.sh"'
             )
+            # We sleep in order to wait for opa to register the constraint
+            # TODO: Is there a better way to do this?
```

Comment:
you might be able to use [the lightkube client.wait() method](https://lightkube.readthedocs.io/en/stable/client/) here instead of this loop. If the resource you are wanting to wait on has a ready condition that should work.

Reply:
The comment was misleading, the problem is we need to wait for the `webhook` to become active. See a related gatekeeper [issue](https://github.com/open-policy-agent/gatekeeper/issues/1156). The gatekeeper helm [hook](https://github.com/open-policy-agent/gatekeeper/blob/master/cmd/build/helmify/static/templates/_helpers.tpl#L51) curls the service until it is up.

The weird thing is that the pebble health check should be enough to ensure when the webhook will be ready and by running `juju status` I can see that the `gatekeeper-controller-manager` app changes from `waiting` to `active` when the webhook is ready and the sleeping stops, but `model.wait_for_idle(status="active")` returns before the application goes `active`. It returns when the units go active instead.

After looking into the `wait_for_idle` code I think that that is the intended behavior, i.e. it checks for the status of the units to determine whether it should return, I guess i will have to find some other way to ensure that the application is `active` or open a PR against juju.